File: src/parsergenerator/nullableset.scala
package pg.parsergenerator

import pg.language.GrammarDefinition
import pg.token.Token

object nullableset {

  /**
   * ある非終端記号から空列が導かれうるかどうかを判定する
   */
  type NullableSet = Set[Token]

  /**
   * @param {GrammarDefinition} grammar 構文規則
   */
  def generateNulls(grammar: GrammarDefinition):NullableSet = {
    // 制約条件を導出するために、
    // 空列になりうる記号の集合nullsを導出
    var nulls = grammar.filter{_.pattern.length==0}.map(_.ltoken).toSet
    // 右辺の記号の数が0の規則を持つ記号は空列になりうる

    // 変更が起きなくなるまでループする
    var flg_changed: Boolean = true
    while (flg_changed) {
      flg_changed = false
      for (rule <- grammar if (!nulls.contains(rule.ltoken) && !rule.pattern.exists(!nulls.contains(_)))) {
        // 既にnullsに含まれていればスキップ
        // 右辺に含まれる記号がすべてnullableの場合はその左辺はnullable
        flg_changed = true
        nulls += rule.ltoken
      }
    }
    nulls
  }

  /**
   * 与えられた[[Token]]がNullableかどうかを調べる
   * @param {Token} token
   * @returns {boolean}
   */
  def isNullable(set:NullableSet, token: Token):Boolean = set.contains(token)
}

File: yacc/nullableset.ml
open Language
open Token
open Utils

(* ある非終端記号から空列が導かれうるかどうかを判定する *)
type nullableSet = S.t

(* 空列になりうる記号の集合 nulls を導出 *)
let generateNulls grammar : nullableSet =
  (* 右辺の記号の数が0の規則を持つ記号は空列になる *)
  let nulls = grammar |>
    List.filter(fun (_,pattern,_)->List.length pattern=0) |>
    List.map(fun (ltoken,_,_)->ltoken) |>
    S.of_list
  in
  (* さらに、空になる文法しか呼び出さない文法要素を追加する *)
  (* 要素を追加したらもう一回調べ直す必要があるのでループ *)
  let rec loop nulls =
    let add_grammer = grammar |> List.filter(fun (ltoken,pattern,_)-> (* 追加するのは *)
      not (S.mem ltoken nulls) && (* nullsに含まれておらず *)
      (* 右辺がすべてnullのもの。つまり、nullにならない要素が含まれないもの *)
      not (List.exists (fun a-> not (S.mem a nulls)) pattern)
    ) in
    if add_grammer = [] then nulls else(* 加えるものがなくなったら終了 *)
    (* 加える文法要素があれば、文法名を取り出し、追加してループします *)
    loop (add_grammer|>(List.fold_left (fun nulls (ltoken,_,_) -> S.add ltoken nulls) nulls))
  in
  loop nulls

(* Token が Nullable かどうか *)
let isNullable nulls token : bool = S.mem token nulls
